pkg/terraform/stages/alibabacloud/stages.go

```diff
@@ -19,4 +19,9 @@ var PlatformStages = []terraform.Stage{
 		[]providers.Provider{providers.AliCloud},
 		stages.WithNormalBootstrapDestroy(),
 	),
+	stages.NewStage(
+		"alibabacloud",
+		"slb_attach_controlplane",
+		[]providers.Provider{providers.AliCloud},
+	),
```

Comment:
Is this handled implicitly by the bootstrap VM being deleted?

Reply:
yes

Reply:
Do you think there is any value in explicitly removing the bootstrap VM from the load balancer? If we are going to rely on implicit behavior, then we at least need some good comments explaining why the bootstrap VM does not need to be explicitly removed from the load balancer. A year from now, I am going to look at this code and not remember why we are not removing from the load balancer: I am going to think it was an oversight.

Reply:
I think there is value in explicitly removing the instance from the SLB. I'm not sure why in 4.10 (older terraform library) release the behavior worked perfectly. The newer terraform libarary no longer functions the same way. If we would prefer to dive into the terraform and figure out why removing the code that removes the bootstrap node also is removing the nodes, we could investigate that. In the meantime, would you prefer if I leave the code commented out in the bootstrap/main.tf with an explanation as to why this occurs? 

I'm fine putting comments in the code. Please specify where you would like them. Suggestions, in the terraform? in the stages? 

I agree that in 6 months from now it will be a vague memory. There might be a fix in the future to how the attachment works in the terraform and we can remove this stage but I believe this should still function when that behavior lands.

Thanks!

Reply:
I was envisioning that the "slb_attach_controlplane" stage would run a new apply after bootstrap complete to redo the `alicloud_slb_backend_server.slb_attach_controlplane` resource so that it would not include the bootstrap VM.

This would be similar to how it works for gcp.
https://github.com/openshift/installer/blob/d5016a08312a971130ab8ea9377da5d5bcc7ec88/data/data/gcp/post-bootstrap/main.tf#L13-L27.

Reply:
This is an interesting approach. The only issue I have is that the behavior should be working fine today but there must be a bug in the tf of the teardown call of removing a backend server from the SLB. Would removing the backend in a separate place (the slb_attachment) behave similarly as the behavior we are seeing now?

Would you like to merge this code and add a follow up PR or would you like to PR something and I can test your suggested changes? I'm fine with either approach. We've pinpointed the issue and now we need to fix. I'm open to whatever approach you would like.

Reply:
It may end up being the same. It is hard to say without understanding why the bug is happening in the terraform provider. There would be a subtle difference in how terraform handles it with the proposed approach.Specifically, the alicloud_slb_backend_server for the bootstrap would be removed via a `terraform apply` of a plan that includes all of the alicloud_slb_backend_servers for the masters as well. This is in contrast to a `terraform destroy` of a plan that only includes the bootstrap one.

Reply:
> We've pinpointed the issue and now we need to fix.

The fix should be in the terraform provider, ultimately.

Reply:
I have a test branch that leaves the teardown of the bootstrap lb connection in the bootstrap/main.tf but then only adds the control plane nodes back inside of the lb attach. This seems to be working. 

https://github.com/kwoodson/installer/tree/lbfix_v1

Reply:
There are no differences between `lbfix` and `lbfix_v1`.

Reply:
Sorry, I pushed the branch without the latest changes. I have pushed the changes.

Reply:
So the effective difference is that the master backends are created after the bootstrap backends, right? If that works, I think that is a cleaner approach. I would like to see some comments on the stage explaining why the master backends need to be added after the bootstrap backends.

I also think that this will be a bit easier to revert once the bug is fixed in the terraform provider.

Reply:
Running into the same issue now. Wondering if there is a race condition upon running the next stage. I've had a successful cluster install and a failure. The bootstrap destroy ran but the slb_attachment stage never ran. 

```
DEBUG Plan: 0 to add, 0 to change, 11 to destroy. 
...
DEBUG Destroy complete! Resources: 11 destroyed.   

<-- loadbalancer backends were removed here -->

DEBUG Loading Install Config...                    
DEBUG   Loading SSH Key...                         
DEBUG   Loading Base Domain...                     
DEBUG     Loading Platform...                      
DEBUG   Loading Cluster Name...                    
DEBUG     Loading Base Domain...                   
DEBUG     Loading Platform...                      
DEBUG   Loading Networking...                      
DEBUG     Loading Platform...                      
DEBUG   Loading Pull Secret...                     
DEBUG   Loading Platform...                        
DEBUG Using Install Config loaded from state file  
INFO Waiting up to 40m0s (until 3:20PM) for the cluster at https://api.test.alicloud-dev.devcluster.openshift.com:6443 to initialize... 
W0316 14:40:59.738715 1143968 reflector.go:324] k8s.io/client-go/tools/watch/informerwatcher.go:146: failed to list *v1.ClusterVersion: Get "https://api.test.alicloud-dev.devcluster.openshift.com:6443/apis/config.openshift.io/v1/clusterversions?fieldSelector=metadata.name%3Dversion&limit=500&resourceVersion=0": dial tcp 47.253.132.94:6443: connect: connection refused
```

@staebler Any ideas why the next stage did not proceed?

Reply:
Do you have a full log that you can show me? I suspect that you are not running the right binary if the "slb_attach_controlplane" stage did not run.

Reply:
After some troubleshooting in cluster.go and watching the stages, I can see that the stages execute sequentially. The cluster, bootstrap, and then the attachment run successfully. Then the teardown occurs for the bootstrap which removes the backend servers from the SLB.

Is there a way to defer the attachment until the teardown of the bootstrap is complete?